Proposed change for Focus Navigation

TakayoshiKochi opened the pull request:
As discussed in
https://github.com/w3c/webcomponents/issues/375

This adds a clarification about focus navigation order for
nodes distributed under a slot element.

Also added handling tabindex="-1" case.

TakayoshiKochi committed:
Proposed change for Focus Navigation

As discussed in
https://github.com/w3c/webcomponents/issues/375

This adds a clarification about focus navigation order for
nodes distributed under a slot element.

Also added handling tabindex="-1" case.

TakayoshiKochi committed:
Proposed change for Focus Navigation

As discussed in
https://github.com/w3c/webcomponents/issues/375

This adds a clarification about focus navigation order for
nodes distributed under a slot element.

Also added handling tabindex="-1" case.

TakayoshiKochi commented:
@hayatoito PTAL

hayatoito commented on spec/shadow/index.html:
```diff
@@ -1139,6 +1141,27 @@
           </li>
         </ol>
 
+        <p>Likewise, the <a>sequential focus navigation order</a> for <a>distributed nodes</a>
```
Wrong... This should not be _distributed nodes_. This should be _assigned nodes_.

In addition to that, you should consider the followings:
1. Do not forget to include descendants of assigned nodes into this focus navigation group.
2. Consider the fallback contents of slots, in addition to assigned nodes.
3. When you try to consider 1 and 2, do not forget the case where a slot is a descendant of another slot.
4. The traversal order is not defined for _sequential focus navigation order for distributed nodes_. Define the traversal order of this focus navigation group.
5. Exclude an assigned node (and more) from its root node's focus navigation group. One node should not join more than one focus navigation group.
6. Mention how tabindex values are _scoped_ at each focus navigation order.

TakayoshiKochi replied:
Thanks.

Probably we want to start using '[control group](https://html.spec.whatwg.org/multipage/interaction.html#control-group)' (defined in HTML spec) for the scope.
I am working on rewriting with it.

hayatoito commented on spec/shadow/index.html:
```diff
@@ -1139,6 +1141,27 @@
+        <p class="note">When a <a>slot</a> <var>A</var> is assigned to another
+          <a>slot</a> <var>B</var>,
+          the <a>sequential focus navigation order</a> of <var>A</var>'s <a>distributed nodes</a>
+          is inserted into <var>B</var>'s <a>sequential focus navigation order</a> following
```
We do not need to mention this.
Please try to have a general rule which covers this case.

hayatoito replied:
Wrong. A's scope is not flattened into B.

hayatoito commented on spec/shadow/index.html:
```diff
@@ -1139,6 +1141,27 @@
+            <a>distributed nodes</a> <strong>must</strong> be inserted in place of
+            <var>A</var> according to its <a>tabindex</a> value.</li>
+        </ol>
+        <p class="note">When a <a>slot</a> <var>A</var> is assigned to another
```
We do not need to mention this. Please try to have a _general_ rule which covers all cases.

Unless a good general rule, the spec would have a lot of cases which are not covered.
e.g.

``` html
<slot>
  <input></input>
  <input>
    <slot>
       <input><input>
     </slot>
  </input>
</slot>
```

hayatoito commented on spec/shadow/index.html:
```diff
@@ -1139,6 +1141,27 @@
+            were assigned the <a>tabindex</a> value 0 for determining its position.</li>
+          <li>Otherwise the <a>sequential focus navigation order</a> for its
+            <a>distributed nodes</a> <strong>must</strong> be inserted in place of
+            <var>A</var> according to its <a>tabindex</a> value.</li>
```
You might want to clarify what "according to it's tabindex value" means.

A slot element is not focusable element. Thus, you should mention how tabindex value of a slot element is _effectively_ used in focus navigation order more clearly.

hayatoito committed:
[Fix #277] Add loadedmetadata to the list

hayatoito committed:
[Fix #373] Make event.deepPath() return an empty array if the event is no longer dispatched

hayatoito committed:
[Fix #386] Fix an editorial issue

hayatoito committed:
Fix an editorial issue

TakayoshiKochi committed:
Updated the draft.

Addresses comments from hayato.

TakayoshiKochi committed:
Updated the draft.

Addresses comments from hayato.

TakayoshiKochi commented:
@hayatoito PTAL.

I tried to get this upstreamable to HTML spec, [section 6.4](https://html.spec.whatwg.org/multipage/interaction.html#focus),
but it is hard to do in a single step, as the HTML spec describes the
navigation order as an algorithm to iteratively find the next focusable node,
while the Shadow DOM spec (before my modification) assumes there is a
(somewhat) pre-computed _focus navigation order_ for each scope and
describes how they are merged together.

So this is my attempt to include the #375 as exactly as possible by
following the current way of describing focus navigation in Shadow DOM.

At this point it seems to me to patch the upstream spec requires lot more work.

hayatoito commented on spec/shadow/index.html:
```diff
@@ -1123,13 +1123,46 @@
         <p>If a <a>node</a> doesn’t <a data-lt="participates">participate</a> in the <a>document flat tree</a>, the <a>node</a> <strong>must</strong> be skipped from the <a>sequential focus navigation</a>.</p>
 
-        <p>The <a>sequential focus navigation order</a> for a given <a>shadow tree</a> <var>A</var> <strong>must</strong> be inserted into the <a>sequential focus navigation order</a> for the <a>parent tree</a> <var>B</var> as follows:</p>
+        <p>Each <a>shadow root</a> and <a>slot</a> element is a <a>control group owner</a> and
```
A control group is more upper layer concept, isn't it?

I mean:
- There is 1 : N relationship between [a control group] and [focus navigation order in each scope].
- There is 1 : 1 relationship between [a control group] and [_global_ focus navigation order], which is the result of _merging_ multiple [focus navigation order in each scope].

TakayoshiKochi replied:
My understanding is that frames own its own control group (e.g. `<iframe>`) and its own sequence,  but from user's standpoint, tab navigation sequence look _global_.  Thus sequences for each group is merged.

hayatoito commented on spec/shadow/index.html:
```diff
@@ -1123,13 +1123,46 @@
+          and descendants of the <a>assigned nodes</a>.
+          If a given slot's assigned nodes is empty, its
+          <a>sequential focus navigation order</a> consists of
+          <a data-lt="focusable area">focusable areas</a> of its fallback contents.
```
Fallback contents are not well defined.
Also do not forget that `<slot>` can be a descendant of another `<slot>`. The current definition does not cover this case at all.

hayatoito commented:
I think you will not get a good definition by the current approach. I recommend you to have a more formal approach:

e.g.

algorithm:
- Input: a document composed tree
- Output: set of owners, where each owner owns a group

algorithm:
- Input: node, NODE, in a document composed tree
- Output: a group which NODE belongs to

Thus, we can define the disjoint sets clearly. Then,

algorithm:
- Input: node A and node B where A nad B joins the same group
- Output: A or B, which comes before in the group

Thus, the _oder_ is clearly defined.

TakayoshiKochi commented:
It looks iterative reviews won't work well if completely different structure is suggested.

I've written the draft following your suggestion:
https://docs.google.com/document/d/1mNV2Z231SrAZuMJvBVNfVNHfULkMwUZOZ75vZM1K9zY/edit?usp=sharing (anyone can comment)

For the 3rd algorithm (ordering) in the comment,
for document/shadow root scope, the order is probably obvious (tree order),
but for slot scope (not fallback content scope) the traversal order is depth-first
search on result of getAssignedNodes()?

In addition, do we want to describe how to compare 2 focusable areas with tabindex here
to get the scoped order of the sequence?

That would add almost dead-copy of [HTML 6.4.3](https://html.spec.whatwg.org/multipage/interaction.html#the-tabindex-attribute) and
super redundant.

TakayoshiKochi commented:
@hayatoito comment on the doc linked above is welcome.

siusin committed:
Used ReSpec

siusin committed:
used ReSpec and added a new WD

siusin committed:
fixed markup errors

hayatoito committed:
Merge pull request #394 from siusin/gh-pages

Get ready for automatic publication - the HTML Imports spec

siusin committed:
fixed broken links

plehegar committed:
Switched custom spec to respec

plehegar committed:
oops, removed respec debug

plehegar committed:
Use default ref for ecmacript

plehegar committed:
Switched DOM ref

plehegar committed:
Improved the respec markup. Moved image closer to the spec.

plehegar committed:
Manifest for automatic publication

plehegar committed:
Fixed broken links per #395

plehegar committed:
Use https

siusin committed:
Merge pull request #395 from siusin/gh-pages

fixed broken links

siusin committed:
merge changes

siusin committed:
Merge branch 'imports/respec' into gh-pages

siusin committed:
Merge pull request #396 from w3c/custom/respec

Switched custom spec to respec

plehegar committed:
Fixed document.registerElement placement

plehegar committed:
Fixed respec markup

hayatoito committed:
editorial: Use experimantal styles

hayatoito committed:
[Fix #402] Use dfn.js for backlinks

hayatoito committed:
[Fix #389] Clarify Window is not shown for the purpose of the explanation

domenic committed:
Large custom element spec rewrite to implement some F2F decisions

Notable changes:

- Implemented HTMLElement constructor using @rniwa's algorithm from #403.
- Rewrote element upgrading to use @rniwa's algorithm from #403, and incorporated it into the rest of the upgrading considerations.
- Got rid of the ability to extend SVGElement, thus allowing us to remove most mentions of namespaces from the spec.
- Removed createdCallback.
- Rewrote "registering elements":
  - Uses defineElement instead of registerElement
  - Preserves the constructor instead of grabbing the .prototype property and synthesizing a new constructor
  - No longer spread out over four separate algorithms plus registerElement; everything is now inline under defineElement
  - More precise about exactly how to get the custom element prototype and callbacks
- Rewrote createElement and createElementNS to be replacements instead of patches, and to call the author-supplied constructor.
- Removed the "All Algorithms in One Diagram" section since so many algorithms changed or were inlined into their callers.
- Removed the "Custom Elements and ECMAScript 6" section since it is very obsolete and does not reflect our current thinking.
- New and rewritten algorithms do not use the unorthodox INPUTS/OUTPUTS blocks, or capitalized variable names. This is kind of a nice marker of new vs. old content.
- I have taken over as spec editor for custom elements

Notable things *not* substantially changed:

- Parser changes are not specced still, besides to say that they should construct the element using its constructor.
- Lifecycle callbacks were not changed, except for removing createdCallback.
- Type extensions were not removed (yet?); it seems better to have a modernized version of them that we atomically remove.
- Registries were not made available everywhere.

Closes #403. Closes #365. Closes #283. Closes #185. Closes #170. Closes #169. Closes #167. Closes #163. Closes #162. Closes #161. Closes #158. Closes #137 (modulo the fact that #165 is still open). Closes #134. Closes #133.

domenic committed:
Add dfn.js to custom elements (part of #402)

domenic committed:
Fix Web IDL spec reference

hayatoito committed:
[Issue #382] Remove component trees, fragment trees, in favor of node trees

hayatoito committed:
[Issue #382] Fix the definition of "in a composed document"

hayatoito committed:
[Fix #410] Clarify the slotting algorithm

hayatoito committed:
Make Flatting Example more readable and understandable

hayatoito committed:
Non-normative: Replace document element with body because document element does not support attachShadow

TakayoshiKochi commented:
@hayatoito
Updated the draft:
https://docs.google.com/document/d/1mNV2Z231SrAZuMJvBVNfVNHfULkMwUZOZ75vZM1K9zY/edit?usp=sharing (anyone can comment)

This should be better aligned with what you suggested... any feedback is welcome.
Sorry for the long turnaround.

hayatoito committed:
[Issue #80] Remove Node.treeRoot because it was renamed and was upstreamed to DOM Standard as Node.rootNode

domenic committed:
Decouple Custom Elements from HTML Imports

Fixes #413. Now HTML Imports patches Custom Elements, instead of Custom
Elements depending on HTML Imports.

domenic committed:
Fix #416: rename "custom element type" to "custom element name"

domenic committed:
Clean up definition of custom element concept

domenic committed:
Fix #369: clarify custom element registries

Every Document now has a custom element registry (no longer just
"registry"). It is null by default. When initializing a new document
during navigation, we give the document an empty registry.

We no longer give template content contents owner documents an empty
registry, which prohibits:

templateEl.contents.ownerDocument.defineElement(...)

And, we no longer have createDocument/createHTMLDocument-created
documents share their creator's registry, which prohibits

document.implementation.createHTMLDocument().defineElement(...)

Finally, if we perform document.open(), we re-set the document's custom
element registry, since all of its elements are now derived from the
"wrong" HTMLElement constructor.

domenic committed:
Send custom element callback errors to "report an exception"

Fixes #174; fixes #421.

domenic committed:
Fix #144: add note about upgrading after cloning

domenic committed:
Update custom element lifecycle

This redefines the custom element callback queue to be a custom element
action queue, to allow upgrades (which are no longer callback-based) to
take part. It also replaces attachedCallback/detachedCallback with
connectedCallback/disconnectedCallback, fixing #362. And it updates all
the places that enqueue these callbacks to be phrased in terms of
precise patches to DOM, except for upgrades.

domenic committed:
Fix typo in createElement

hayatoito committed:
[Fix #61] Add a note that event section is meaningful only if an event target is a Node

domenic committed:
Remove document argument to (dis)connectedCallback

Re-closes #362. See https://github.com/w3c/webcomponents/issues/362#issuecomment-193695492.

domenic committed:
Factor out "create an element" algorithm

This fixes a few bugs and edge cases in createElement(NS). The eventual
goal, per #424, is to have all element creation go through this.

domenic committed:
Use new "create an element" hook in cloning and parsing

Tentatively closes #424.

domenic committed:
Change createElement(ln, typeExtension) to createElement(ln, { is })

Fixes #408.

domenic committed:
Enqueue attributeChangedCallback during parsing

Fixes #364. Also s/attributes list/attribute list/g.

domenic committed:
Remove leftover warning

The relevant text was removed in
674b0b2b75b380d036ca2aab3ba423c9d13f0799.

domenic committed:
Implement observedAttributes processing

Fixes #367. As discussed there "observedAttributes" is a better name for
an opt-in list than "attributeFilter" (which sounds like opt-out).

domenic committed:
Define valid custom element name more rigorously

This defines a valid custom element name with a grammar, as is done for
the XML definitions that createElement(NS) uses. This closes #239 by
taking the following fairly liberal intersection:

- Allowed in createElement
- Allowed in createElementNS
- Initial [a-z] required for HTML parser compatibility
- Hyphen required
- Uppercase characters disallowed
- Blacklist disallowed

Also, clean up how the spec talks about the two types of custom elements
(custom tags and type extensions).

domenic committed:
Overhaul custom element upgrading to be much more explicit

This eliminates a lot of action-at-a-distance from the spec, in favor of
having the appropriate steps set or unset the new "defined flag". This
fixes #425. We also make the element interface concept more explicit,
with a new section stating the exact text that should be added to HTML.

While doing this, also  changed :unresolved to :defined, which fixes
#418. You can use `:not(:defined)` for the effect of the old
`:unresolved`.

The term "resolved" has been eliminated from the spec in favor of
"defined".

domenic committed:
Make monkeypatching more uniform and explicit

domenic committed:
Don't default extends/is to null

Per @annevk's comments in https://github.com/w3c/webcomponents/commit/0abc10e43ac827b593db326a74bf1af9ee7ced37.

domenic committed:
Add note explaining custom element name restrictions

Fixes #166.

domenic committed:
Explain type extensions better in a non-normative note

Closes #147.

hayatoito committed:
[Fix #393] Remove the statement of cloneNode on ShadowRoot

hayatoito committed:
[Fix #429] Integrate "get the parent" model for an event path

domenic committed:
Do not hold a reference to all undefined potentially-custom elements

This is part of #419. There remains work to be done for figuring out
more ergonomic ways of making after-the-fact upgrade easy; see
https://github.com/w3c/webcomponents/issues/419#issuecomment-194640804.

domenic committed:
Dial back auto-upgrading of undefined potentially-custom elements

Per https://github.com/w3c/webcomponents/issues/419#issuecomment-194984109,
there is a preference for the fragment parser and cloning algorithm to
not create elements that can be auto-upgraded.

domenic committed:
Typo fix

domenic committed:
Specify custom elements parser behavior

Fixes #366. Fixes #165. Fixes #151.

domenic committed:
Touchup definition of "list of defined custom element local names"

domenic committed:
Flesh out defineElement examples

domenic committed:
Add checks for non-callable lifecycle callbacks at registration time

domenic committed:
Add conformance requirements for custom element constructors

Fixes #409.

domenic committed:
Allow early return in conformance requirements

domenic committed:
Note some subtleties around constructor vs. attachedCallback

See https://github.com/w3c/webcomponents/issues/409#issuecomment-195119284.

domenic committed:
attached/detached → connected/disconnected

Somehow my brain hadn't made the switch so every time I worked on new
text I kept reintroducing the old name. Oops.

domenic committed:
Remove some semicolons

Operating under the assumption that this is going to DOM, @annevk's
specs don't use semicolons :-S

hayatoito committed:
[Fix #433] Clarify that relatedTarget property is non-null

domenic committed:
Mention that SVG and MathML won't get new hyphen-containing names either

Fixes #436.

domenic committed:
Add a reference to XML's EBNF notation

Fixes #435.

domenic committed:
Reorganize custom elements, including API changes for definition

- Figure out where custom elements will go in HTML, and started writing as if they were there
- Fleshes out and moved around examples and motivation to form an "introduction" section
- Separates out and rearranges concepts a bit better (fixes #437)
- Moves the registry to the Window instead of the Document (fixes #369)
- Moves to `window.customElements`, an instance of `CustomElementsRegistry`, with one `define` method currently (fixes #431)

domenic committed:
Fix typo in API summary for customElements.define

TakayoshiKochi commented:
@hayatoito Updated the draft. PTAL.

hayatoito committed:
[Issue #441] Add a node about ShadowRoot.innerHTML

hayatoito committed:
[Fix #91] Use DocumentOrShadowRoot Mixin

hayatoito committed:
[Issue #91] Clarify elementFromPoint should use retargeting algorithm

hayatoito committed:
[Fix #443] Clarify innerHTML should work the same way

Commit:
Update from the draft document.

- markup for <var>
- period or colon at the end of bullet
- more links to definition
- remove title.
- fix autolink

Commit:
Update from the draft document.

- markup for <var>
- period or colon at the end of bullet
- more links to definition
- remove title.
- fix autolink

domenic committed:
Update custom element examples to use new define function

Re-fixes #431. Closes #446. Also updates acknowledgments to be easier to
maintain, and adds @rniwa who was missing (in addition to @rianby64 who
spotted this and proposed #446).

Commit:
Merge branch 'focusScope' of github.com:TakayoshiKochi/webcomponents into focusScope

TakayoshiKochi commented:
@hayatoito this branch went into some weird state, so I created another PR #447.
Could you take a look?
